Implement updater for Windows and Linux

Ryex opened the pull request:
branch rename of #1097

Requires #1102 and #1227

Close #94 
Close #167

As I previously discussed on the discord the options for auto updating on windows are abysmal. we are frankly better off rolling our own (simple) solution.

## The Plan

- [x] Pull release info from the github api
- [x] Check if there is a newer release
- [x] Ask the user if they wish to update
- [x] Launch update process and exit prism.
- [x] (Optional) Let the user select from available newer releases
- [x] Download appropriate release asset to temp location (accounting from architecture, portable install etc. Simple so long as our naming remains consistent) (make the algo fuzy to account for variance if it's is needed)
- [x] create local file that marks an update in progress
- [x] if there are files we **know** need to be removed between updates "remove" them but create back-up (os move)
- [x] perform update
  - [x] IF PORTABLE: 
    - [x] Extract to temp local folder
    - [x] create maker file in tmp folder so updater knows what's going on (contains target install path)
    - [x] Run new updater, exit current updater process
    - [x] new updater process moves files out of temp dir to install dir and copies itself over old updater then exits 
  - [x] IF MSI / Setup.exe INSTALL:
    - [x] Run installer and exit updater
  - [x] IF Appimage 
    - [x] attempt to call embedded `AppImageUpdate`
- [x] Next time prism starts is sees marker files, present user with update log and cleans up.

Presumably at every point in the process if there is an error there is still an intact launcher install

Ryex committed:
feat: paliminary updater

- can check for need to update
- can select a version to update to
- perform update: TODO

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix: typo - no space before comment

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
Revert "fix: typo - no space before comment"

This reverts commit 0b2044e9a434ba3d2724630399bc41ebb321b039.

Ryex committed:
fix: typo - space before comment

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
Merge branch 'packaging/windows-installer-uninstall-previous' into feat/windows-updater

Ryex committed:
packaging: include file manifest in portable install

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
packaging: ensure all files are in manifest

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
packaging: use powershell name

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
packaging: trim manfest paths to be relative

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
packaging: msys2 shell is bash

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
packaging: add manifest.txt to linux tarballs

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
packaging: trim leading working dir from manifest paths

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
packaging: convert msys paths

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
packaging: fix typo ; in middle of line

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
refactor: rename updater files

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
refactor: rename app class

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
refactor: split out setting api headers for downloads

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
Merge branch 'refactor/net-split-headers-to-proxy-class' into feat/windows-updater

Ryex committed:
Merge branch 'packaging/windows-installer-uninstall-previous' into feat/windows-updater

Ryex committed:
Merge branch 'packaging/file-manifest-in-portable-install' into feat/windows-updater

Ryex committed:
refactor: use Net tasks for github api download

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
feat(updater) select valid asset

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
packaging: add appimage update capability

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
packaging(linux): use vars when refrencing qt install dir

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
packaging(appimage): dont use rsync

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
packaging: use runner.workspace when copying files

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
packaging(appimage): put zsync in relase:x

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
Merge branch 'packaging/appimage-updates' into feat/launcher-updater

Ryex committed:
feat(updater): download new & back up old

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
feat(updater): unpack portable archive

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
support windows console

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix: filter archive assets from windows non-portable installs

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
feat(updater): final step for portable install

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
feat(updater): packaging - know the updater git repo

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
DCO Remediation Commit for Rachel Powers <508861+Ryex@users.noreply.github.com>

I, Rachel Powers <508861+Ryex@users.noreply.github.com>, hereby add my Signed-off-by to this commit: f5729d9a7c953d88ae60e106a787b1d7c35ede8b

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
Merge branch 'packaging/appimage-updates' into feat/launcher-updater

Ryex committed:
chore: add license to headers

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
Merge branch 'refactor/net-split-headers-to-proxy-class' into feat/launcher-updater

Ryex committed:
fix: use new shared pointer for repsonse buffer

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
Merge branch 'refactor/net-split-headers-to-proxy-class' into feat/launcher-updater

Ryex committed:
fix: include `^<optional^>` updater doesn't need to know about mods

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
Merge branch 'refactor/net-split-headers-to-proxy-class' into feat/launcher-updater

Ryex committed:
feat(updater): tie in updater part 1

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
Proper capture on windows

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
feat(updater): tie in part 2, let there be UI!

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix(updater): logs/ folder

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Scrumplex commented on .github/workflows/build.yml:
```diff
@@ -275,12 +283,12 @@ jobs:
         shell: msys2 {0}
         run: |
-          cmake -S . -B ${{ env.BUILD_DIR }} -DCMAKE_INSTALL_PREFIX=${{ env.INSTALL_DIR }} -DCMAKE_BUILD_TYPE=${{ inputs.build_type }} -DENABLE_LTO=ON -DLauncher_BUILD_PLATFORM=${{ matrix.name }} -DCMAKE_C_COMPILER_LAUNCHER=${{ env.CCACHE_VAR }} -DCMAKE_CXX_COMPILER_LAUNCHER=${{ env.CCACHE_VAR }} -DLauncher_QT_VERSION_MAJOR=6 -DCMAKE_OBJDUMP=/mingw64/bin/objdump.exe -G Ninja
+          cmake -S . -B ${{ env.BUILD_DIR }} -DCMAKE_INSTALL_PREFIX=${{ env.INSTALL_DIR }} -DCMAKE_BUILD_TYPE=${{ inputs.build_type }} -DENABLE_LTO=ON -DLauncher_BUILD_PLATFORM=${{ matrix.name }} -DCMAKE_C_COMPILER_LAUNCHER=${{ env.CCACHE_VAR }} -DCMAKE_CXX_COMPILER_LAUNCHER=${{ env.CCACHE_VAR }} -DLauncher_QT_VERSION_MAJOR=6 -DCMAKE_OBJDUMP=/mingw64/bin/objdump.exe -DLauncher_UPDATER_GITHUB_REPO ...
```
let's move this into our CMakeLists.txt

Everything else (i.e. macOS updater) is also in there anyway

Scrumplex commented on .github/workflows/build.yml:
```diff
@@ -275,12 +283,12 @@ jobs:
         if: runner.os == 'Windows' && matrix.msystem == ''
         run: |
-          cmake -S . -B ${{ env.BUILD_DIR }} -DCMAKE_INSTALL_PREFIX=${{ env.INSTALL_DIR }} -DCMAKE_BUILD_TYPE=${{ inputs.build_type }} -DENABLE_LTO=ON -DLauncher_BUILD_PLATFORM=${{ matrix.name }} -DLauncher_QT_VERSION_MAJOR=${{ matrix.qt_ver }} -DCMAKE_MSVC_RUNTIME_LIBRARY="MultiThreadedDLL" -A${{ matrix.architecture}} -DLauncher_FORCE_BUNDLED_LIBS=ON
+          cmake -S . -B ${{ env.BUILD_DIR }} -DCMAKE_INSTALL_PREFIX=${{ env.INSTALL_DIR }} -DCMAKE_BUILD_TYPE=${{ inputs.build_type }} -DENABLE_LTO=ON -DLauncher_BUILD_PLATFORM=${{ matrix.name }} -DLauncher_QT_VERSION_MAJOR=${{ matrix.qt_ver }} -DCMAKE_MSVC_RUNTIME_LIBRARY="MultiThreadedDLL" -A${{ matrix.architecture}} -DLauncher_FORCE_BUNDLED_LIBS=ON -DLauncher_UPDATER_GITHUB_REPO=https://github ...
```
ditto

Scrumplex commented on .github/workflows/build.yml:
```diff
@@ -295,7 +303,7 @@ jobs:
         if: runner.os == 'Linux'
         run: |
-          cmake -S . -B ${{ env.BUILD_DIR }} -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_BUILD_TYPE=${{ inputs.build_type }} -DENABLE_LTO=ON -DLauncher_BUILD_PLATFORM=Linux -DCMAKE_C_COMPILER_LAUNCHER=${{ env.CCACHE_VAR }} -DCMAKE_CXX_COMPILER_LAUNCHER=${{ env.CCACHE_VAR }} -DLauncher_QT_VERSION_MAJOR=${{ matrix.qt_ver }} -G Ninja
+          cmake -S . -B ${{ env.BUILD_DIR }} -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_BUILD_TYPE=${{ inputs.build_type }} -DENABLE_LTO=ON -DLauncher_BUILD_PLATFORM=Linux -DCMAKE_C_COMPILER_LAUNCHER=${{ env.CCACHE_VAR }} -DCMAKE_CXX_COMPILER_LAUNCHER=${{ env.CCACHE_VAR }} -DLauncher_QT_VERSION_MAJOR=${{ matrix.qt_ver }} -DLauncher_UPDATER_GITHUB_REPO=https://github.com/${{ github.repository }} -G Ninja
```
ditto

DioEgizio commented on .github/workflows/build.yml:
```diff
@@ -383,12 +391,13 @@ jobs:
         if: runner.os == 'Windows' && matrix.msystem == ''
         run: |
           cmake --install ${{ env.BUILD_DIR }} --config ${{ inputs.build_type }}
+          Get-Content ${{ env.BUILD_DIR }}/install_manifest.txt | %{ $_.TrimStart("$pwd/") } | %{ $_.TrimStart('${{ env.INSTALL_DIR }}') } | %{ $_.TrimStart('./') } | Out-File -Append -FilePath ${{ env.INSTALL_DIR }}/manifest.txt 
```
Why is it doing it twice???

Ryex replied:
merge artifact. I'll mark it for removal

DioEgizio reviewed (commented):
Don't we also need to sign prismlauncher_updater too now?

Ryex commented:
> Don't we also need to sign prismlauncher_updater too now?

we do, thanks for reminding me.

Ryex committed:
fix: prep for changes in #1276

- sign updater

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
add optional pre-release tag

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
no need for pre release tag

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix: add thread sleep to wait for resources - add detail text from logs

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
Merge branch 'develop' into feat/launcher-updater

Ryex committed:
fix: add split markdown source

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
Merge branch 'refactor/net-split-headers-to-proxy-class' into feat/launcher-updater

Ryex committed:
fix add NetRequest source

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix: add build config header & trimlines form stdout

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix: final fixes

- use `done(code)` for offer dialog

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix: copy needs to overwrite

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix(updater) fixes form first round of testing

- reset update time after check

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fixa(updater): better update timer logs

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

flowln committed:
fix(updater): add parent to dialog to fix issues in tilling WMs

QProgressDialog without parents in Qt tend to cause graphical issues on
some systems, so we add the main window as a parent here!

Signed-off-by: flow <flowlnlnln@gmail.com>

Ryex committed:
fix: adjust dialog size, add parent to msgboxes

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix(updater): convert int to string

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix(updater): set minimum dialog / msgbox sizes

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix(updater): Explicit conversion to string for QByteArray

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix(updater): include updater in setup.exe

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

DioEgizio commented on launcher/Application.cpp:
```diff
@@ -462,48 +567,48 @@ Application::Application(int &argc, char **argv) : QApplication(argc, argv)
+        qDebug() << "Compiled for               : " << BuildConfig.systemID();
+        qDebug() << "Compiled by                : " << BuildConfig.compilerID();
+        qDebug() << "Build Artifact             : " << BuildConfig.BUILD_ARTIFACT;
+        qDebug() << "Updates Enabeled           : " << (updaterEnabled() ? "Yes" : "No");
```
```suggestion
        qDebug() << "Updates Enabled           : " << (updaterEnabled() ? "Yes" : "No");
```

Ryex committed:
Update launcher/Application.cpp

Co-authored-by: DioEgizio <83089242+DioEgizio@users.noreply.github.com>
Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix(updater): avoid windows installer detection with "asInvoker" in xml exe.manifest

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix(updater): force `asInvoker` for updater on windows
  - use `__COMPAT_LAYER` env var in windows to bypass Installer Detection and run as a normal user anyway
  - set up updater directly after main windows is created
  - check update locks before main window

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix(updater): check the app root dir for binary

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix(updater): build atrifact fix on linux ci + add qt-ver to artifact name

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix(updater): ensure updater knows binaries are in `./bin/` on linux and can find them

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix(updater): fix bad exe_name on linux

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix(windows installer): old installers didn't uninstall first.

now they do so ensure shortcuts stay selected by default

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
Merge branch 'develop' into feat/launcher-updater

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
Merge branch 'fix/windows-console-attach' into feat/launcher-updater

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

# Conflicts:
#	launcher/Application.cpp
#	launcher/CMakeLists.txt
#	launcher/filelink/FileLink.cpp

Ryex committed:
fix: religate export zip to launcher app only

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

fix(MMCZip): include QUrl

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

fix(FileLink): drop FreeConsole

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

fix(filelink exe): add console sources

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix(flame creation task): import ApiDownload

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
refactor(filelink): drop moved funciton

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Trial97 committed:
Merge branch 'develop' of https://github.com/PrismLauncher/PrismLauncher into feat/launcher-updater

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

TheKodeToad committed:
Merge branch 'develop' into feat/launcher-updater

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

Trial97 committed:
Merge branch 'develop' of https://github.com/PrismLauncher/PrismLauncher into feat/launcher-updater

Trial97 committed:
format

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 commented:
~~Fixed the AppImageUpdate stuff, if I remember correctly windows update works.
So taking all this into consideration this should be ready for review, and maybe we organize a testing on the weekend for this.~~
Let me know if I missed anything. 
I did miss a lot of a stuff: 
- [x] appimageupdate is inside the appimage
- [x] and did not touch the portable stuff

Trial97 committed:
fixed appImageUpdate

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
more fixing

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
updated portable update

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
chaged folder again

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 commented:
Now I'm pretty sure this is ready for review 
Working for me:
- linux portable update
- appimage update

I did not touch any windows related code so that one I expect it to run as before.

DioEgizio commented:
> Now I'm pretty sure this is ready for review Working for me:
> 
>     * linux portable update
> 
>     * appimage update
> 
> 
> I did not touch any windows related code so that one I expect it to run as before.

i think we should have another test in #testing to see if everything works fine, just in case

Trial97 committed:
Merge branch 'develop' of https://github.com/PrismLauncher/PrismLauncher into feat/launcher-updater

Trial97 committed:
fixed updater build

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
Merge branch 'develop' of https://github.com/PrismLauncher/PrismLauncher into feat/launcher-updater

Trial97 committed:
added explicit qt version for linux portable

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

RokeJulianLockhart commented:
This won't activate in managed versions like the Flatpak package, right?

Trial97 commented:
No Flatpack package should not be affected.
